backend/app/services/openalex.py

```
@@ -288,3 +295,109 @@ async def fetch_related_earlier_papers(self, paper: dict, limit: int = 20) -> li
+                score += query_overlap * 1.5
+                score += min(log10((candidate.get("citedByCount") or 0) + 1), 4.0) * 0.35
+                if candidate.get("year") is not None:
+                    score += max(0, (year - candidate["year"])) * 0.01
+
+                existing = merged.get(candidate_id)
+                if existing is None or score > existing[0]:
+                    merged[candidate_id] = (score, candidate)
+
+        ranked: list[dict] = []
+        for _, candidate in sorted(merged.values(), key=lambda item: item[0], reverse=True):
+            title_key = _normalize_text(candidate.get("title", ""))
```

Ruff (0.15.11)

[error] Undefined name `_normalize_text`

(F821)